spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -4005,6 +4012,62 @@ private Long computeBackwardWhereTo(long offset, boolean toCurrent, TopicPartiti
 
 		}
 
+		private void invokeBatchWithIndividualRecordObservation(final ConsumerRecords<K, V> records,
+				List<ConsumerRecord<K, V>> recordList, @Nullable Object sample) {
+
+			List<Observation> observations = new ArrayList<>();
+
+			try {
+				// Create individual observations for each record without scopes
+				for (ConsumerRecord<K, V> record : recordList) {
+					Observation observation = KafkaListenerObservation.LISTENER_OBSERVATION.observation(
+							this.containerProperties.getObservationConvention(),
+								DefaultKafkaListenerObservationConvention.INSTANCE,
+								() -> new KafkaRecordReceiverContext(record, getListenerId(), getClientId(), this.consumerGroupId,
+										this::clusterId),
+								this.observationRegistry);
+					observation.start();
+					observations.add(observation);
+				}
+
+				// Invoke the batch listener
+				if (this.wantsFullRecords) {
+					Objects.requireNonNull(this.batchListener).onMessage(records,
+							this.isAnyManualAck
+									? new ConsumerBatchAcknowledgment(records, recordList)
+									: null,
+							this.consumer);
+				}
+				else {
+					doInvokeBatchOnMessage(records, recordList);
+				}
+
+				batchInterceptAfter(records, null);
+				successTimer(sample, null);
+			}
+			catch (RuntimeException e) {
+				this.batchFailed = true;
+				failureTimer(sample, null, e);
+				batchInterceptAfter(records, e);
+
+				// Mark all observations with error
+				for (Observation observation : observations) {
+					if (!isListenerAdapterObservationAware()) {
+						observation.error(e);
+					}
+				}
+				throw e;
+			}
+			finally {
```

Comment:
if we switch to `Observe.observe()`, then there is no need for this `finally` block as `observe()` will implicitly do that for you.